File: Design Patterns/Behavioral/Observer.ts
// used for Real Updates 
interface Observer { update(data: any): void; }


interface Subject {
  attach(observer: Observer): void;
  detach(observer: Observer): void;
  notify(): void;
}


class ConcreteSubject implements Subject {
  private observers: Observer[] = [];
  private state: string = "";

  attach(observer: Observer): void { this.observers.push(observer); }

  detach(observer: Observer): void { this.observers = this.observers.filter(obs => obs !== observer); }

  setState(state: string): void {
    this.state = state;
    this.notify();
  }

  getState(): string { return this.state; }

  notify(): void { this.observers.forEach(observer => observer.update(this.state)); }
}


class ConcreteObserver implements Observer {
  constructor(private name: string) { }

  update(data: any): void { console.log(`${this.name} received update: ${data}`); }
}


const subject = new ConcreteSubject();

const observer1 = new ConcreteObserver("Observer 1");
const observer2 = new ConcreteObserver("Observer 2");

subject.attach(observer1);
subject.attach(observer2);

subject.setState("New State!");

File: Design Patterns/Behavioral/C#/Observer.cs
using System;
using System.Collections.Generic;

interface IObserver
{
  void Update(object data);
}

interface ISubject
{
  void Attach(IObserver observer);
  void Detach(IObserver observer);
  void Notify();
}

class Subject : ISubject
{
  private List<IObserver> observers = new List<IObserver>();
  private string state = "";
  public void Attach(IObserver observer) { this.observers.Add(observer); }

  public void Detach(IObserver observer) { this.observers.RemoveAll(obs => obs != observer); }
  public void SetState(string state)
  {
    this.state = state;
    this.Notify();
  }
  public string GetState() { return this.state; }

  public void Notify()
  {
    foreach (var observer in this.observers)
      observer.Update(this.state);
  }
}

class Observer : IObserver
{
  private string name;

  public Observer(string name) { this.name = name; }

  public void Update(object data) { Console.WriteLine($"{this.name} received update: {data}"); }
}

class Program
{
  public static void Main(string[] args)
  {
    Subject subject = new Subject();
    Observer observer1 = new Observer("Observer 1");
    Observer observer2 = new Observer("Observer 2");

    subject.Attach(observer1);
    subject.Attach(observer2);
    subject.SetState("New State!");
  }
}